## Семинар 3. Изменение таблиц в Pandas

## Условие 1: Задача 1
- Скачать данные по ссылке https://www.kaggle.com/datasets/ionaskel/laptop-prices
- Считать данные с помощью pandas
- Вывести на экран первые 5 строк

In [114]:
import pandas as pd

In [115]:
df = pd.read_csv('kc_house_data.csv', parse_dates=['date'], dayfirst=False)
df.head()

id       date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520 2014-10-13  221900.0         3       1.00         1180   
1  6414100192 2014-12-09  538000.0         3       2.25         2570   
2  5631500400 2015-02-25  180000.0         2       1.00          770   
3  2487200875 2014-12-09  604000.0         4       3.00         1960   
4  1954400510 2015-02-18  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### 1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома
Если реновации дома не было, то в новом признаке поставьте 0


In [116]:
delta = []
for i, row in df.iterrows():
  if row['yr_renovated']:
      delta.append(row['yr_renovated'] - row['yr_built'])
  else:
      delta.append(0)
df['delta_renovated'] = delta
df.sample(16)

id       date      price  bedrooms  bathrooms  sqft_living  \
5285   2896600020 2015-03-25   460000.0         3       1.75         1520   
4711   8081900195 2014-12-10   572500.0         3       1.00         1590   
17172  9238450430 2014-06-24   275000.0         3       1.00          990   
15068  7202330920 2014-08-07   464000.0         3       2.50         1690   
10169  8570900162 2014-10-16   193500.0         2       1.00          950   
2807   8857640860 2014-09-26   522000.0         4       2.50         2835   
20320  7203140420 2015-01-28   385000.0         3       2.50         1740   
13448  2426059124 2014-12-16  1045000.0         4       3.25         4160   
15214  3876500290 2015-03-05   175000.0         3       1.00         1070   
14831  1843130360 2015-05-07   295000.0         3       2.50         2030   
11897  8944600060 2015-04-20   541900.0         3       2.50         1880   
20822  3204930510 2015-02-24   780000.0         5       3.50         3190   
6861   5381000048 2015-04-27   110000.0         2       1.00          790   
6869   5104510130 2014-06-03   312000.0         4       2.50         1830   
5280   3856904970 2014-08-18   469000.0         2       1.00         1120   
345    3969300030 2014-07-23   165000.0         4       1.00         1000   

       sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
5285       7700     1.0           0     0  ...         820            700   
4711       4600     1.5           0     0  ...        1290            300   
17172      9798     1.0           0     0  ...         990              0   
15068      4898     2.0           0     0  ...        1690              0   
10169     15996     1.0           0     0  ...         950              0   
2807       6598     2.0           0     0  ...        2835              0   
20320      4145     2.0           0     0  ...        1740              0   
13448     47480     2.0           0     0  ...        4160              0   
15214      6164     1.0           0     0  ...        1070              0   
14831      4867     2.0           0     0  ...        2030              0   
11897      3054     2.0           0     0  ...        1880              0   
20822      4247     2.0           0     0  ...        2430            760   
6861       8250     1.0           0     0  ...         790              0   
6869       5175     2.0           0     0  ...        1830              0   
5280       4284     1.0           0     0  ...         730            390   
345        7134     1.0           0     0  ...        1000              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
5285       1969             0    98034  47.7226 -122.219           1420   
4711       1926             0    98117  47.6807 -122.399           1770   
17172      1968             0    98072  47.7670 -122.164           1210   
15068      2003             0    98053  47.6834 -122.038           2220   
10169      1946          1995    98045  47.4987 -121.787            950   
2807       2002             0    98038  47.3878 -122.034           2770   
20320      2010             0    98053  47.6875 -122.015           1740   
13448      1995             0    98072  47.7266 -122.115           3400   
15214      1967             0    98001  47.3377 -122.291           1320   
14831      2003             0    98042  47.3747 -122.128           2030   
11897      1988             0    98007  47.6091 -122.147           1840   
20822      2013             0    98052  47.7016 -122.103           2580   
6861       1947             0    98188  47.4523 -122.286            900   
6869       2003             0    98038  47.3565 -122.016           1830   
5280       1921             0    98105  47.6688 -122.324           2050   
345        1943             0    98178  47.4897 -122.240           1020   

       sqft_lot15  delta_renovated  
5285         7674                0  
4711         4350                0  
17172   

### 1.3 Создайте признаки года продажи, месяца продажи


In [117]:
df['yr_sold'] = df['date'].dt.year
df['month_sold'] = df['date'].dt.month_name()
df.sample(10)

id       date      price  bedrooms  bathrooms  sqft_living  \
13729  1423049029 2015-03-06   265000.0         4       1.75         1970   
5306   8087800020 2014-05-15   412500.0         3       1.50         1490   
4120    952004875 2014-06-02   661000.0         4       2.25         1990   
17305   452002135 2015-04-22  1070000.0         4       2.50         2740   
1823   4322500055 2015-04-17   607000.0         5       1.75         1910   
11111  8856920550 2015-03-19   378000.0         3       2.50         2130   
1297   2492200435 2014-06-06   389250.0         2       1.50         1490   
636    6131600060 2014-08-15   214000.0         3       1.00         1200   
10857  2310030390 2015-04-29   275000.0         3       1.75         1180   
11911  9828700685 2015-05-13   900000.0         3       1.75         1540   

       sqft_lot  floors  waterfront  view  ...  yr_built  yr_renovated  \
13729      8390     1.0           0     0  ...      1955             0   
5306       8475     1.0           0     0  ...      1963             0   
4120       4600     1.5           0     2  ...      1932             0   
17305      5000     2.0           0     0  ...      2012             0   
1823       5428     1.0           0     0  ...      1954             0   
11111      8404     2.0           0     0  ...      1991             0   
1297       4080     1.0           0     0  ...      1956             0   
636        8316     1.0           0     0  ...      1953             0   
10857      6260     1.0           0     0  ...      1993             0   
11911      8400     1.5           0     0  ...      1902             0   

       zipcode      lat     long  sqft_living15  sqft_lot15  delta_renovated  \
13729    98178  47.4861 -122.251           1710       10890                0   
5306     98052  47.6571 -122.133           1490        8540                0   
4120     98126  47.5638 -122.380           1810        5750                0   
17305    98107  47.6740 -122.371           1660        5000                0   
1823     98136  47.5333 -122.392           1820        5900                0   
11111    98058  47.4623 -122.130           2130        8404                0   
1297     98126  47.5344 -122.381           1320        4080                0   
636      98002  47.3221 -122.215           1200        8316                0   
10857    98038  47.3531 -122.048           1580        7272                0   
11911    98122  47.6187 -122.295           1330        4800                0   

       yr_sold  month_sold  
13729     2015       March  
5306      2014         May  
4120      2014        June  
17305     2015       April  
1823      2015       April  
11111     2015       March  
1297      2014        June  
636       2014      August  
10857     2015       April  
11911     2015         May  

[10 rows x 24 columns]

### 1.4 Удалите признаки date, zipcode, lat, long

In [118]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
df.sample(10)

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
8987   5035300650  1099500.0         4       2.00         2580      6000   
16604  1432700420   250000.0         3       1.00         1460     10914   
4174   2461900945   438000.0         5       1.00         1950      6250   
7418   3881900321   339950.0         3       1.00         1050      5402   
13722  4039700870  1160000.0         5       1.75         2870      9680   
13674  6163900333   338000.0         3       1.75         1250      7710   
2556   2973800030   175000.0         3       1.00         1030      6600   
2039   3395040550   320000.0         3       2.50         1530      2890   
14877  3274850130   392000.0         4       2.50         2600      8921   
3145   9274201807   580000.0         3       2.50         1590      1937   

       floors  waterfront  view  condition  grade  sqft_above  sqft_basement  \
8987      1.0           0     0          5      9        1300           1280   
16604     1.0           0     0          4      7        1460              0   
4174      1.5           0     0          4      7        1450            500   
7418      1.5           0     0          4      7        1050              0   
13722     1.0           0     4          5      9        1440           1430   
13674     1.0           0     0          4      7        1250              0   
2556      1.0           0     0          3      7        1030              0   
2039      2.0           0     0          3      7        1530              0   
14877     2.0           0     0          3      8        2600              0   
3145      2.5           0     0          3      8        1590              0   

       yr_built  yr_renovated  sqft_living15  sqft_lot15  delta_renovated  \
8987       1950             0           2300        6200                0   
16604      1959             0           1490        8314                0   
4174       1917             0           1950        6250                0   
7418       1906             0           1510        4685                0   
13722      1966             0           2940        9729                0   
13674      1947             0           1340        7710                0   
2556       1954             0           1220        9040                0   
2039       2001             0           1540        2890                0   
14877      1991             0           2550        8683                0   
3145       2004             0           1620        2022                0   

       yr_sold month_sold  
8987      2014       June  
16604     2014       July  
4174      2015      April  
7418      2014        May  
13722     2015   February  
13674     2014   November  
2556      2014       July  
2039      2015      April  
14877     2014   December  
3145      2014     August

## Условие 2: 2 задача
Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})


In [119]:
clients = pd.DataFrame({
    'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295, 939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242, 2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
    'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207, 6821600300, 7137950720, 9510920050, 6131600255, 5428000070, 1788800910, 8100400160, 3123049142, 6306800010, 5083000375,
                 920100025, 1951600150, 809001400, 339600110, 1622049154, 1099600250, 8563000110, 2768100205, 3995700435, 8861700030, 3303980210, 7731100066, 8146100580, 825069097, 3889100029, 9524100196]
})
clients.head()

client_id    house_id
0       1459  8965450190
1       4684  6823100225
2       3498  5104540330
3       3942  2131701075
4       4535  1522700060

## 2.1 Присоедините к таблице clients данные по домам через метод join


In [120]:
cl_id = clients.set_index('house_id')
cl_id.head()

client_id
house_id             
8965450190       1459
6823100225       4684
5104540330       3498
2131701075       3942
1522700060       4535

In [121]:
df_id = df.set_index('id')
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          3      7        1180              0      1955   
1           0     0          3      7        2170            400      1951   
2           0     0          3      6         770              0      1933   
3           0     0          5      7        1050            910      1965   
4           0     0          3      8        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  delta_renovated  yr_sold  \
0             0           1340        5650                0     2014   
1          1991           1690        7639               40     2014   
2             0           2720        8062                0     2015   
3             0           1360        5000                0     2014   
4             0           1800        7503                0     2015   

  month_sold  
0    October  
1   December  
2   February  
3   December  
4   February

In [122]:
cl_id.join(df_id).sample(10)

client_id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
3123049142        280  452000.0       3.0       2.25       2600.0   14810.0   
6823100225       4684  700000.0       4.0       1.75       1870.0    6000.0   
8965450190       1459  295000.0       3.0       2.50       1500.0    3060.0   
8563000110       2068  427000.0       4.0       1.75       1460.0    9750.0   
5083000375       4332  170000.0       3.0       1.00       1310.0    9529.0   
809001400        3113  925000.0       3.0       1.00       1630.0    3200.0   
9524100196       1474  239000.0       2.0       1.50        680.0     772.0   
1522700060       4535  518000.0       4.0       2.75       2520.0   14021.0   
920100025        3475       NaN       NaN        NaN          NaN       NaN   
5428000070       2295  770000.0       5.0       3.50       4750.0    8234.0   

            floors  waterfront  view  condition  grade  sqft_above  \
3123049142     1.0         0.0   2.0        4.0    8.0      1490.0   
6823100225     1.0         0.0   0.0        5.0    8.0      1670.0   
8965450190     2.0         0.0   0.0        3.0    7.0      1500.0   
8563000110     1.0         0.0   0.0        4.0    7.0      1460.0   
5083000375     1.0         0.0   0.0        3.0    7.0      1310.0   
809001400      1.5         0.0   0.0        4.0    8.0      1630.0   
9524100196     2.0         0.0   0.0        3.0    7.0       680.0   
1522700060     2.0         0.0   0.0        3.0    9.0      2520.0   
920100025      NaN         NaN   NaN        NaN    NaN         NaN   
5428000070     2.0         0.0   2.0        3.0   10.0      3350.0   

            sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
3123049142         1110.0    1956.0           0.0         2450.0     16715.0   
6823100225          200.0    1949.0           0.0         1710.0      6000.0   
8965450190            0.0    1994.0           0.0         2700.0      7734.0   
8563000110            0.0    1967.0           0.0         1820.0      9840.0   
5083000375            0.0    1956.0           0.0         1330.0      9529.0   
809001400             0.0    1912.0           0.0         1710.0      3600.0   
9524100196            0.0    2005.0           0.0          690.0      1059.0   
1522700060            0.0    1999.0           0.0         2330.0     14007.0   
920100025             NaN       NaN           NaN            NaN         NaN   
5428000070         1400.0    2013.0           0.0         2160.0     14496.0   

            delta_renovated  yr_sold month_sold  
3123049142              0.0   2014.0     August  
6823100225              0.0   2015.0      April  
8965450190              0.0   2015.0   February  
8563000110              0.0   2015.0      April  
5083000375              0.0   2014.0    October  
809001400               0.0   2014.0  September  
9524100196              0.0   2014.0   November  
1522700060              0.0   2014.0       June  
920100025               NaN      NaN        NaN  
5428000070              0.0   2015.0        May

## 2.2 Присоедините к таблице clients данные по домам через метод merge. Это нужно, чтобы понимать, какие дома покупались клиентами. house_id - это индексы датафрейма с домами

In [123]:
clients.merge(df, left_on='house_id', right_on='id').head()

client_id    house_id          id     price  bedrooms  bathrooms  \
0       1459  8965450190  8965450190  295000.0         3       2.50   
1       4684  6823100225  6823100225  700000.0         4       1.75   
2       3498  5104540330  5104540330  679000.0         4       2.50   
3       3942  2131701075  2131701075  420000.0         3       1.75   
4       4535  1522700060  1522700060  518000.0         4       2.75   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1500      3060     2.0           0  ...      7        1500   
1         1870      6000     1.0           0  ...      8        1670   
2         3680      7236     2.0           0  ...     10        3680   
3         1720      5000     1.5           0  ...      8        1720   
4         2520     14021     2.0           0  ...      9        2520   

   sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
0              0      1994             0           2700        7734   
1            200      1949             0           1710        6000   
2              0      2006             0           3310        7180   
3              0      1932          2009           1410        8300   
4              0      1999             0           2330       14007   

   delta_renovated  yr_sold  month_sold  
0                0     2015    February  
1                0     2015       April  
2                0     2015         May  
3               77     2014    December  
4                0     2014        June  

[5 rows x 22 columns]

## Условие 3: 3 задача - Составьте несколько сводных таблиц
### 3.1 Найдите среднюю стоимость домов в зависимости от количества спален. Отсортируйте от меньшей стоимости к большей


In [124]:
df.groupby('bedrooms').agg({'price': 'mean'}).sort_values('price').round(2)

price
bedrooms            
1          317642.88
2          401372.68
0          409503.85
3          466232.08
11         520000.00
4          635419.50
33         640000.00
5          786599.83
10         819333.33
6          825520.64
9          893999.83
7          951184.66
8         1105076.92

### 3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома


In [125]:
df.groupby('condition').agg({'price': ['min', 'mean', 'max']}).round(2)

price                      
                min       mean        max
condition                                
1           78000.0  334431.67  1500000.0
2           80000.0  327287.15  2555000.0
3           75000.0  542012.58  7062500.0
4           89000.0  521200.39  7700000.0
5          110000.0  612418.09  3650000.0

### 3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида


In [126]:
df.pivot_table(index='waterfront', columns='view', values='id', aggfunc='count', fill_value=0)

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

In [127]:
df.groupby(['view', 'waterfront']).agg({'id': 'count'}).unstack('view').fillna(0)

id                            
view              0      1      2      3      4
waterfront                                     
0           19489.0  331.0  955.0  491.0  184.0
1               0.0    1.0    8.0   19.0  135.0

### 3.4 Каких домов в зависимости от этажности и количества спален больше?


In [128]:
pd.crosstab(index=df['floors'], columns=df['bedrooms'])

bedrooms  0    1     2     3     4    5    6   7   8   9   10  11  33
floors                                                               
1.0        4  162  1951  5455  2383  605  104   9   5   0   1   0   1
1.5        0   21   182   786   698  185   30   7   1   0   0   0   0
2.0        6   12   497  3118  3682  775  119  19   6   4   2   1   0
2.5        0    1     5    56    58   23   14   2   0   2   0   0   0
3.0        2    3   123   405    61   13    5   1   0   0   0   0   0
3.5        1    0     2     4     0    0    0   0   1   0   0   0   0

Домов с 1 этажом и 3 спальнями больше (5455).

### 3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома

In [129]:
df.pivot_table(index='condition', columns='grade', values='price', aggfunc='median', fill_value=0)

grade          1       3       4       5       6       7       8        9   \
condition                                                                    
1          142000       0  150000  190000  255000  403500  932500        0   
2               0  280000  145000  180000  235000  305000  429000   715000   
3               0   75000  205000  234475  265000  357500  485000   689000   
4               0       0  238525  229975  295000  390000  571250   823500   
5               0  262000       0  227450  285475  456000  696000  1078000   

grade           10       11       12       13  
condition                                      
1                0        0        0        0  
2          1752500        0        0        0  
3           890000  1209500  1807500  2888000  
4          1030000  1685000  2125000  5750000  
5          1650000  2050000  1990000        0

In [130]:
df.groupby(['condition', 'grade']).agg({'price': 'median'}).unstack('grade').fillna(0)

price                                                    \
grade            1         3         4         5         6         7    
condition                                                               
1          142000.0       0.0  150000.0  190000.0  255000.0  403500.0   
2               0.0  280000.0  145000.0  180000.0  235000.0  305000.0   
3               0.0   75000.0  205000.0  234475.0  265000.0  357500.0   
4               0.0       0.0  238525.0  229975.0  295000.0  390000.0   
5               0.0  262000.0       0.0  227450.0  285475.0  456000.0   

                                                                            
grade            8          9          10         11         12         13  
condition                                                                   
1          932500.0        0.0        0.0        0.0        0.0        0.0  
2          429000.0   715000.0  1752500.0        0.0        0.0        0.0  
3          485000.0   689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4          571250.0   823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          696000.0  1078000.0  1650000.0  2050000.0  1990000.0        0.0